### PR TITLE
Add Interval classes

### PR DESCRIPTION
**Pull request by danvk**
New modules:

- `Interval` represents a closed numeric interval.
- `ContigInterval` represents a closed interval on the genome (i.e. a contig + an `Interval`).
- `utils` contains a few helpers for ordering tuples, ala Python.

All of these are well-tested. None of them are used right now, but they'll all be helpful for working with bigBed files.

### CODE REVIEW

**Comment by danvk** [comment]
If there are existing libraries for doing all this, I'd be happy to use them instead! A quick search on NPM didn't turn up anything quite like what I wanted—I think my `Interval` class is too simple for people to consider packaging it up.

File: src/ContigInterval.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class ContigInterval {
+  contig: string|number;
+  interval: Interval;
+
+  constructor(contig: string|number, start: number, stop: number) {
+    this.contig = contig;
+    this.interval = new Interval(start, stop);
+  }
+
+  // TODO: make these getter methods & switch to Babel.
```
Comment (ihodes) [inline src/ContigInterval.js]: Sat more? (Not in the code, just curious!)

File: src/ContigInterval.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+var Interval = require('./Interval');
+
+/**
+ * Class representing a closed interval on the genome: contig:start-stop.
+ *
+ * The contig may be either a string ("chr22") or a number (in case the contigs
+ * are indexed, for example).
+ */
+class ContigInterval {
+  contig: string|number;
```
Comment (ihodes) [inline src/ContigInterval.js]: Would this make future code easier to deal with if contigs are always strings?

File: src/ContigInterval.js (hunk 1 above)
Comment (danvk) [inline src/ContigInterval.js]: According to the [es6features repo][1], this is supposed to be valid syntax:

```javascript
class Interval {
  get length() {
    return this.stop - this.start + 1;
  }
}
```

and then you can access `length` as though it were a property, not a method:

```javascript
new Interval(0, 10).length
```

I believe this is possible in ES5 using `Object.defineProperty`, but I don't want to muck with that. `react-tools` specifically doesn't support that syntax, but Traceur & Babel do.

[1]: https://github.com/lukehoban/es6features#classes

Comment (ihodes) [inline src/ContigInterval.js]: Ah yeah, cool, makes sense. That's definitely possible in ES5 & it prevents the JIT from kicking, in just FYI.

File: src/ContigInterval.js (hunk 2 above)
Comment (danvk) [inline src/ContigInterval.js]: In bigBed format, they use a mapping which converts all the contig names to integers, for more compact storage. So it is convenient to allow `contig` to be either.

File: src/Interval.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    var result = intervals[0].clone();
+    intervals.slice(1).forEach(function({start, stop}) {
+      result.start = Math.max(start, result.start);
+      result.stop = Math.min(stop, result.stop);
+    });
+    return result;
+  }
+
+  static boundingInterval(intervals: Array<Interval>): Interval {
```
Comment (ihodes) [inline src/Interval.js]: Took me a minute to figure out what this was doing; maybe a comment?

Comment (danvk) [inline src/Interval.js]: Done

File: src/ContigInterval.js (hunk 1 above)
Comment (danvk) [inline src/ContigInterval.js]: > it prevents the JIT from kicking

as in it's bad for performance?

Comment (ihodes) [inline src/ContigInterval.js]: Depending on how often it's called, it could be.

**Comment by ihodes** [comment]
Cool, LGTM!

**Pull request merged**